### DISCUSSION TITLE
Sloeber Install / Installation advice

### DISCUSSION THREAD

**Discussion by bukwyrm** [Sloeber/arduino-eclipse-plugin#1549] [Q&A]
Hi, i had Sloeber running on Win7, but am now trying for Ubuntu 22 LTS, and running into some problems. 

I'll start off with some ideas on how to make the experience more fun (Please don't take this the wrong way, i am both thankful and in awe of the development effort that went into this neat program, just some hints on the user experience from view of a noob)

- it would be nice to have one consistent tenor on eclipse.baeyens.it, consolidating the scattered java info:  
  /installAdvice.shtml : "**java -showversion It should give you 1.8. (higher versions will not work)**"
  /stable.php?OS=Linux : "**Sloeber needs java 17 or higher but is now included in the release!**"
  /how_to.shtml#/d1 : "**Make sure you have Java 8 on your system.**"
- a link to this here github page on the "Community : Forum Discussion" item would be great, instead of the link to the Arduino forum-thread. Or both?

Now to my actual Q:

1. I have Ubuntu 22.04 LTS 64bit, 
2. downloaded V4.4.1_linux64.2022-07-09_05-48-31.zip, 
3. did all the sudo stuff on /installAdvice.shtml-Linux, 
4. unpacked the archive, 
5. opened a terminal, went to /Downloads/V4.4.1_linux64.2022-07-09_05-48-31/Sloeber/, 
6. did chmod u+x sloeber-ide , then did "Start Sloeber" by entering ./sloeber-ide in the terminal. 
7. I then got a popup-window that had the following info: 

```
JVM terminated. Exit code=13
/home/xyz/Downloads/V4.4.1_linux64.2022-07-09_05-48-31/Sloeber//plugins/org.eclipse.justj.openjdk.hotspot.jre.full.linux.x86_64_17.0.3.v20220515-1416/jre/bin/java
-Dosgi.requiredJavaVersion=11
-Dorg.eclipse.update.reconcile=false
-Dlog4j.configuration=log4j/log4j.xml
-Xms256m
-Xmx1024m
-jar /home/xyz/Downloads/V4.4.1_linux64.2022-07-09_05-48-31/Sloeber//plugins/org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar
-os linux
-ws gtk
-arch x86_64
-showsplash
-launcher /home/xyz/Downloads/V4.4.1_linux64.2022-07-09_05-48-31/Sloeber/sloeber-ide
-name Sloeber-ide
--launcher.library /home/xyz/Downloads/V4.4.1_linux64.2022-07-09_05-48-31/Sloeber//plugins/org.eclipse.equinox.launcher.gtk.linux.x86_64_1.2.500.v20220509-0833/eclipse_11701.so
-startup /home/xyz/Downloads/V4.4.1_linux64.2022-07-09_05-48-31/Sloeber//plugins/org.eclipse.equinox.launcher_1.6.400.v20210924-0641.jar
--launcher.appendVmargs
-exitdata 8017
```

I checked my installed java with java -showversion and got (although it should not matter, as sloeber inlcudes the java, right?) 
```
openjdk version "17.0.5" 2022-10-18
OpenJDK Runtime Environment (build 17.0.5+8-Ubuntu-2ubuntu122.04)
OpenJDK 64-Bit Server VM (build 17.0.5+8-Ubuntu-2ubuntu122.04, mixed mode, sharing)
```

The stuff i found on JVM exit code 13 was mostly a possible incompatibility of 32/64bit versions, but my system is 64, my sloeber dl is 64, my java is 64, and i also did the i386 stuff shown on the sloeber linux comment page

Can someone point me to a solution or ask some crucial details?

**Comment by jantje** [comment]
I agree the website is getting pretty outdated :-(. Also my build engine doesn't run anymore so the copy from github to the website is manual :-(

But what I think is the missing step:
Java is included by a plugin that downloads java (when sloeber starts)
But due to security the downloaded java needs to be chomd +x
That is why http://eclipse.baeyens.it/stable.php?OS=Linux states

Sloeber needs java 17 or higher but is now included in the release!

Due to new security a simple download and extract will not work.

After the download run.

chmod u+x sloeber-ide

Start Sloeber; then run
chmod +x plugins/org.eclipse.justj.openjdk.hotspot.jre.full.linux.x86_64_*/jre/bin/*

**Reply by bukwyrm** [reply]
Ahh, i think i got it - can it be that the instructions have the steps switched? I now did:

1. downloaded V4.4.1_linux64.2022-07-09_05-48-31.zip,
2. did all the sudo stuff on http://eclipse.baeyens.it/installAdvice.shtml-Linux,
3. unpacked the archive,
4. opened a terminal, went to `/Downloads/V4.4.1_linux64.2022-07-09_05-48-31/Sloeber/`,
5. in this terminal, entered `chmod u+x sloeber-ide` 
6. in this terminal, started Sloeber by entering `./sloeber-ide` (gives error-popup)
7. in this terminal, entered `chmod +x plugins/org.eclipse.justj.openjdk.hotspot.jre.full.linux.x86_64_*/jre/bin/*`
8. in this terminal, started Sloeber by entering `./sloeber-ide` 

Sloeber started, but there were some errors:
```
SLF4J: Failed to load class "org.slf4j.impl.StaticLoggerBinder".
SLF4J: Defaulting to no-operation (NOP) logger implementation
SLF4J: See http://www.slf4j.org/codes.html#StaticLoggerBinder for further details.
SWT SessionManagerDBus: Failed to connect to org.gnome.SessionManager: Failed to execute child process “dbus-launch” (No such file or directory)
SWT SessionManagerDBus: Failed to connect to org.xfce.SessionManager: Failed to execute child process “dbus-launch” (No such file or directory)
start installing platform Arduino AVR Boards avr(1.8.6)

(Sloeber:29659): Gdk-CRITICAL **: 15:04:45.903: gdk_threads_set_lock_functions: assertion 'gdk_threads_lock == NULL && gdk_threads_unlock == NULL' failed
SWT SessionManagerDBus: Failed to connect to org.gnome.SessionManager: Failed to execute child process “dbus-launch” (No such file or directory)
SWT SessionManagerDBus: Failed to connect to org.xfce.SessionManager: Failed to execute child process “dbus-launch” (No such file or directory)

```

**Comment by jantje** [comment]
downloaded V4.4.1_linux64.2022-07-09_05-48-31.zip,
    did all the sudo stuff on http://eclipse.baeyens.it/installAdvice.shtml-Linux,
    unpacked the archive,
    opened a terminal, went to /Downloads/V4.4.1_linux64.2022-07-09_05-48-31/Sloeber/,
    there, entered chmod u+x sloeber-ide
  there, started Sloeber by entering ./sloeber-ide
    there, entered chmod +x plugins/org.eclipse.justj.openjdk.hotspot.jre.full.linux.x86_64_*/jre/bin/*
    there, started Sloeber by entering ./sloeber-ide

**Reply by bukwyrm** [reply]
Many thanks for your continued goodwill - sadly i was not able to install yet. My current protocol:
1. downloaded V4.4.1_linux64.2022-07-09_05-48-31.zip,
2. did all the sudo stuff on http://eclipse.baeyens.it/installAdvice.shtml (->Linux comments),
3. unpacked the archive to `/Downloads/Sloeber/`,
4. opened a terminal on location `/Downloads/Sloeber/`,
5. in this terminal, entered `chmod u+x sloeber-ide` 
6. in this terminal, started Sloeber by entering `./sloeber-ide` (gives error-popup "Exit code = 13", I closed error popup)
7. in this terminal, entered `chmod +x plugins/org.eclipse.justj.openjdk.hotspot.jre.full.linux.x86_64_*/jre/bin/*`
8. in this terminal, started Sloeber by entering `./sloeber-ide` (Sloeber IDE opens, does some stuff, then gives error popup "Failed to install Arduino AVR Boards")

In the (error-) log of the sloeber workspace, the details of the errors are shown as such:
`
```
eclipse.buildId=unknown
java.version=17.0.3
java.vendor=Eclipse Adoptium
BootLoader constants: OS=linux, ARCH=x86_64, WS=gtk, NL=en_US
Framework arguments:  -perspective io.sloeber.application.perspective
Command-line arguments:  -os linux -ws gtk -arch x86_64 -perspective io.sloeber.application.perspective

io.sloeber.core
Error
Thu Jan 19 20:49:02 CET 2023
Failed to install Arduino AVR Boards

eclipse.buildId=unknown
java.version=17.0.3
java.vendor=Eclipse Adoptium
BootLoader constants: OS=linux, ARCH=x86_64, WS=gtk, NL=en_US
Framework arguments:  -perspective io.sloeber.application.perspective
Command-line arguments:  -os linux -ws gtk -arch x86_64 -perspective io.sloeber.application.perspective

io.sloeber.core
Error
Thu Jan 19 20:49:02 CET 2023
Failed to extract tar.bz2. "/home/xyz/Downloads/Sloeber/arduinoPlugin/downloads/avr-1.8.6.tar.bz2".

java.io.IOException: Cannot run program "chmod": error=13, Permission denied
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1143)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1073)
	at java.base/java.lang.Runtime.exec(Runtime.java:594)
	at io.sloeber.core.tools.PackageManager.chmod(PackageManager.java:336)
	at io.sloeber.core.tools.PackageManager.extract(PackageManager.java:259)
	at io.sloeber.core.tools.PackageManager.processArchive(PackageManager.java:84)
	at io.sloeber.core.tools.PackageManager.downloadAndInstall(PackageManager.java:74)
	at io.sloeber.core.tools.PackageManager.downloadAndInstall(PackageManager.java:465)
	at io.sloeber.core.api.BoardsManager.install(BoardsManager.java:230)
	at io.sloeber.core.api.BoardsManager.startup_Pluging(BoardsManager.java:704)
	at io.sloeber.core.Activator$2.run(Activator.java:254)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.io.IOException: error=13, Permission denied
	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
	at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:314)
	at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:244)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1110)
	... 11 more

eclipse.buildId=unknown
java.version=17.0.3
java.vendor=Eclipse Adoptium
BootLoader constants: OS=linux, ARCH=x86_64, WS=gtk, NL=en_US
Framework arguments:  -perspective io.sloeber.application.perspective
Command-line arguments:  -os linux -ws gtk -arch x86_64 -perspective io.sloeber.application.perspective

org.eclipse.urischeme
Error
Thu Jan 19 20:47:59 CET 2023
Cannot run program "xdg-mime": error=13, Permission denied

java.io.IOException: Cannot run program "xdg-mime": error=13, Permission denied
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1143)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1073)
	at org.eclipse.urischeme.internal.registration.ProcessExecutor.execute(ProcessExecutor.java:36)
	at org.eclipse.urischeme.internal.registration.RegistrationLinux.getRegisteredDesktopFileForScheme(RegistrationLinux.java:145)
	at org.eclipse.urischeme.internal.registration.RegistrationLinux.determineHandlerLocation(RegistrationLinux.java:87)
	at org.eclipse.urischeme.internal.registration.RegistrationLinux.getSchemesInformation(RegistrationLinux.java:76)
	at org.eclipse.urischeme.AutoRegisterSchemeHandlersJob.run(AutoRegisterSchemeHandlersJob.java:87)
	at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
Caused by: java.io.IOException: error=13, Permission denied
	at java.base/java.lang.ProcessImpl.forkAndExec(Native Method)
	at java.base/java.lang.ProcessImpl.<init>(ProcessImpl.java:314)
	at java.base/java.lang.ProcessImpl.start(ProcessImpl.java:244)
	at java.base/java.lang.ProcessBuilder.start(ProcessBuilder.java:1110)
	... 7 more
```

`

**Comment by jantje** [comment]
>java.io.IOException: Cannot run program "chmod": error=13, Permission denied

Sloeber needs full read write access to the arduinoplugin folder.  Your setup clearly does not allow Sloeber full access.
You can move the arduinoPlugin folder with the environment variable SLOEBER_HOME

**Reply by dk7ob** [reply]
This is a quite old thread, but I am running into exact the same issue while trying to install 4.4.1, while exactly following the instructions. I moved the  SLOEBER_HOME to different locations, even to /tmp, but always get the same error message. Is it possible to see where the chmod failes (which file or directory)?

The arduinoplugin folder is there, filled with lots of files and the installation already unpacked quite a lot of the files in the downloads subfolder, but stops with "avr-1.8.6.tar.bz2".

Can I create a detailed log of the installation process (apart from strace :-) )?

This is on Ubuntu 22.04.

**Reply by jantje** [reply]
It may be that avr-1.8.6.tar.bz2 in the downloads folder is bad.
It is a standard archive file so any archive file tool should be able to check the file.
When Sloeber needs to download a file it first looks in the downloads folder. If Sloeber finds the file it needs in the downloads folder it will not try to download it again.
So if a download only partially succeeded the installation will continue to fail if you do not manually remove the bad file.
The download folder is only used for caching so you can delete the downloads folder to safe disk space and solve bad download files.

**Reply by dk7ob** [reply]
The download file seems to be OK, I can extract it with tar manually.
I did 2 more things:
1. Replaced the file with the avr-1.8.3 version in my 4.4 installation, renamed it to avr-1.8.6. Same error.
2. Copied the complete arduinoplugin folder from the 4.4 installation. Error is gone, I can build my project, but upload fails with 
`Failed to run the Upload recipe 
Cannot run program "/home/wofritz/Sloeber/arduinoPlugin/packages/arduino/tools/avrdude/6.3.0-arduino17/bin/avrdude" (in directory "/home/wofritz/Documents/sloeber-workspace"): error=13, Keine Berechtigung`

If I copy the avrdude command from the Eclipse console into a shell, it is working!

What else can I try?

**Reply by jantje** [reply]
It does look like there are access right issues on your system. However it is always hard to find what rights are missing.
Given the very little info you give it is even harder to find out.
FI I do not know whether you ran the upload command in directory "/home/wofritz/Documents/sloeber-workspace"
Maybe you should create a issue and we go from there (this is way beyond the discussion)

**Reply by dk7ob** [reply]
***** Start edit 
While trying to create the issue, I found issue #1543 which describes the same problem on a Linux Mint system. The workaround  mentioned there "sudo chmod -R 777 *" in the sloeber installation directory helped and the installation finished without errors. I think I tried something like this before, but without sudo.
***** end edit *****

I started the download in "/home/wofritz/Documents/sloeber-workspace" and it worked.

Please keep in mind that I know nothing about the inner workings of Eclipse and Eclipse plugins. It's a tool for me. I am willing to help but need some guidance.

I am now creating an issue.
